Objects/listobject.c

```diff
@@ -569,24 +566,35 @@ list_repeat(PyListObject *a, Py_ssize_t n)
     if (np == NULL)
         return NULL;
 
+    PyObject **dest = np->ob_item;
+    PyObject **dest_end = dest + size;
     if (Py_SIZE(a) == 1) {
-        items = np->ob_item;
-        elem = a->ob_item[0];
-        for (i = 0; i < n; i++) {
-            items[i] = elem;
-            Py_INCREF(elem);
+        PyObject *elem = a->ob_item[0];
+        Py_SET_REFCNT(elem, Py_REFCNT(elem) + n);
+#ifdef Py_REF_DEBUG
+        _Py_RefTotal += n;
+#endif
+        while (dest < dest_end) {
+            *dest++ = elem;
         }
     }
     else {
-        p = np->ob_item;
-        items = a->ob_item;
-        for (i = 0; i < n; i++) {
-            for (j = 0; j < Py_SIZE(a); j++) {
-                *p = items[j];
-                Py_INCREF(*p);
-                p++;
+        PyObject **src0 = a->ob_item;
+        PyObject **src_end = src0 + Py_SIZE(a);
+        PyObject **src;
+        for (src = src0; src < src_end; src++) {
+            Py_SET_REFCNT(*src, Py_REFCNT(*src) + n);
+#ifdef Py_REF_DEBUG
+            _Py_RefTotal += n;
+#endif
+        }
+        while (dest < dest_end) {
+            src = src0;
+            while (src < src_end) {
+                *dest++ = *src++;
```

Comment:
Have you tried replacing this (& the similar tuple copying loop) with a `memcpy()` call? That would at least be another adventure in cross-platform mysteries :wink:.